Sources/swift-format/Subcommands/DumpConfiguration.swift

```diff
@@ -15,32 +15,55 @@ import Foundation
 import SwiftFormat
 
 extension SwiftFormatCommand {
-  /// Dumps the tool's default configuration in JSON format to standard output.
+  /// Dumps the tool's configuration in JSON format to standard output.
   struct DumpConfiguration: ParsableCommand {
     static var configuration = CommandConfiguration(
-      abstract: "Dump the default configuration in JSON format to standard output"
+      abstract: "Dump the configuration in JSON format to standard output",
+      discussion: """
+        Without any options, dumps the default configuration. When '--effective' is set, dumps the configuration that \
+        would be used if swift-format was executed from the current working directory (cwd), incorporating \
+        configuration files found in the cwd or its parents, or input from the '--configuration' option.
+        """
     )
 
+    /// Whether or not to dump the effective configuration.
+    @Flag(name: .shortAndLong, help: "Dump the effective instead of the default configuration.")
+    var effective: Bool = false
+
+    @OptionGroup()
+    var configurationOptions: ConfigurationOptions
+
+    func validate() throws {
+      if configurationOptions.configuration != nil && !effective {
+        throw ValidationError("'--configuration' is only valid in combination with '--effective'")
+      }
+    }
+
     func run() throws {
-      let configuration = Configuration()
-      do {
-        let encoder = JSONEncoder()
-        encoder.outputFormatting = [.prettyPrinted]
-        if #available(macOS 10.13, *) {
-          encoder.outputFormatting.insert(.sortedKeys)
-        }
-
-        let data = try encoder.encode(configuration)
-        guard let jsonString = String(data: data, encoding: .utf8) else {
-          // This should never happen, but let's make sure we fail more gracefully than crashing, just
-          // in case.
-          throw FormatError(
-            message: "Could not dump the default configuration: the JSON was not valid UTF-8"
-          )
-        }
-        print(jsonString)
-      } catch {
-        throw FormatError(message: "Could not dump the default configuration: \(error)")
+      if !effective {
+        let configuration = try Configuration().asJsonString()
+        print(configuration)
+        return
+      }
+
+      // Pretend to use stdin, so that the configuration loading machinery in the Frontend base class can be used in the
+      // next step. This produces the same results as if "format" or "lint" subcommands were called.
+      let lintFormatOptions = try LintFormatOptions.parse(["-"])
+
+      let frontend = DumpConfigurationFrontend(
+        configurationOptions: configurationOptions,
+        lintFormatOptions: lintFormatOptions
+      )
+      frontend.run()
```

Comment:
I’m not a big fan of running a dummy frontend that stores the configuration used as a side effect. I’d prefer to just call into `Frontend.configuration(fromPathOrString:orInfoerredFromSwiftFileAt:)` instead.

Reply:
Yes, looks like it shouldn’t be too hard. It looks like `ConfigurationLoader` can be a local variable instead of a member and `DiagnsoticsEngine` can be passed in.